### PR TITLE
feat(stageExecution): supporting extra custom tags for stage execution metrics

### PR DESCRIPTION
**Pull request by zrashwani**
Adding the ability to add extra tags to the metrics recorded on the stage execution level. 
This will add flexibility in instrumenting custom stages provided via Orca plugins as follows:

* Adding public methods to set/get metric tags in `StageExecution` interface
* Injecting those custom tags to the metrics recorded in related stage handlers (e.g: `StartStageHandler`, `CompleteStageHandler`)

### CODE REVIEW

File: orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandler.kt [hunk 1]
```diff
@@ -179,6 +179,10 @@ class CompleteStageHandler(
         stage.context["cloudProvider"]?.let {
           id.withTag("cloudProvider", it.toString())
         } ?: id
+      }.let { id ->
+        stage.metricTags?.let {
+          id.withTags(stage.metricTags)
```
Comment (deverton) [inline orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandler.kt:184]: Does this merge or replace the existing tags?

Comment (zrashwani) [reply orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandler.kt:184]: it merge the tags provided as a parameter with existing ones (here is the [implementation of `withTags`](https://github.com/Netflix/spectator/blob/main/spectator-api/src/main/java/com/netflix/spectator/api/DefaultId.java#L71-L73) method)

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/StageExecutionImpl.java [hunk 2]
```diff
@@ -343,6 +343,23 @@ public void setLastModified(@Nullable LastModifiedDetails lastModified) {
     this.lastModified = lastModified;
   }
 
+  /**
+   * Additional tags to be used with stage metrics. This is useful to add extra dimensions to the
+   * metrics recorded for built-in or custom stages.
+   */
+  private Map<String, String> metricTags;
```
Comment (dbyron-sf) [inline orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/StageExecutionImpl.java:350]: a bit of a nit, but perhaps
```suggestion
  private Map<String, String> additionalMetricTags;
```
to match the description?

File: orca-api/src/main/java/com/netflix/spinnaker/orca/api/pipeline/models/StageExecution.java [hunk 3]
```diff
@@ -123,6 +123,11 @@ public interface StageExecution {
 
   void setLastModified(@Nullable LastModifiedDetails lastModified);
 
+  @Nullable
+  Map<String, String> getMetricTags();
+
+  void setMetricTags(@Nullable Map<String, String> metricTags);
```
Comment (dbyron-sf) [inline orca-api/src/main/java/com/netflix/spinnaker/orca/api/pipeline/models/StageExecution.java:129]: We don't need to allow a null argument here, do we?

File: orca-queue/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 4]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (dbyron-sf) [inline orca-queue/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: Sorry, I'm not familliar with this. Why do we need it?

Comment (zrashwani) [reply orca-queue/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: this give us the ability to mock or spy final classes, in this case, I needed to `spy(NoopRegistry)` in the tests [here](https://github.com/spinnaker/orca/pull/4255/files#diff-5030240d38e2011d634b6cf4ab9ef2868655238b0cb1f254c30ba0e4e3f03618R108)

File: orca-api/src/main/java/com/netflix/spinnaker/orca/api/pipeline/models/StageExecution.java (hunk 3 above)
Comment (zrashwani) [reply orca-api/src/main/java/com/netflix/spinnaker/orca/api/pipeline/models/StageExecution.java]: yes, removed nullable from  method parameter

File: orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/StageExecutionImpl.java (hunk 2 above)
Comment (zrashwani) [reply orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/StageExecutionImpl.java]: done

File: orca-queue/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker (hunk 4 above)
Comment (dbyron-sf) [reply orca-queue/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: Thanks!

File: orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandlerTest.kt [hunk 5]
```diff
@@ -101,7 +104,8 @@ object CompleteStageHandlerTest : SubjectSpek<CompleteStageHandler>({
   val publisher: ApplicationEventPublisher = mock()
   val exceptionHandler: ExceptionHandler = DefaultExceptionHandler()
   val clock = fixedClock()
-  val registry = NoopRegistry()
+  val metricId: Id = mock()
+  val registry = spy(NoopRegistry())
```
Comment (dbyron-sf) [inline orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandlerTest.kt:108]: Actually, as I look at this more closely, would it be simpler to change this to DefaultRegistry?  I think there's no longer a need to mock metricId.  The assertions change to something like (untested):
```
    assertThat(registry.timers()).hasSize(1);
    Timer timer = registry.timers().findFirst().orElseThrow(AssertionError::new);
    assertThat(timer.id().name()).isEqualTo("stage.invocations.duration");
    assertThat(timer.id().tags())
        .contains(
            tag("tag1", "value1"),
            tag("tag2", "value2"));
```

Comment (zrashwani) [reply orca-queue/src/test/kotlin/com/netflix/spinnaker/orca/q/handler/CompleteStageHandlerTest.kt]: Yea, using DefaultRegistry does simplify things. I have modified that and removed the metricId mock, however in a different way from the snippet above, as `timers` have more than one element, and we cannot assume it has size=1.
I modified to validate the timer using `argumentCaptor` as below:
```
            argumentCaptor<Id>().let {
              verify(registry, atLeastOnce()).timer(it.capture())
              val metricId = it.firstValue
              assertThat(metricId.name()).isEqualTo("stage.invocations.duration")
              assertThat(metricId.tags().any{ t -> t.key() == "tag1" && t.value() == "value1" }).isTrue
              assertThat(metricId.tags().any{ t -> t.key() == "tag2" && t.value() == "value2" }).isTrue
            }
```

File: orca-queue/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker (hunk 4 above)
Comment (dbyron-sf) [inline orca-queue/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: I think this can disappear now.

Comment (zrashwani) [reply orca-queue/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: no, it is still needed, as `spy(DefaultRegistry())` is used in the tests, and DefaultRegistry is also a final class

Comment (dbyron-sf) [reply orca-queue/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: Aaah, ok.

**Review by dbyron-sf — APPROVED** [review]

**Pull request merged**